src/charm.py

```diff
@@ -16,13 +21,213 @@ class GraylogCharm(CharmBase):
+
+        # event observations
         self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.update_status, self._on_update_status)
+        self.framework.observe(self.on.stop, self._on_stop)
+        self.framework.observe(
+            self.on['elasticsearch'].relation_changed,
+            self._on_elasticsearch_relation_changed
+        )
+        self.framework.observe(
+            self.on['elasticsearch'].relation_broken,
+            self._on_elasticsearch_relation_broken
+        )
+        self.framework.observe(
+            self.on["mongodb"].relation_changed, self._on_mongodb_relation_changed
+        )
+        self.framework.observe(
+            self.on["mongodb"].relation_broken, self._on_mongodb_relation_broken
+        )
+        # initialized stored variables
+        self._stored.set_default(elasticsearch_uri=str())  # connection str for elasticsearch
+        self._stored.set_default(mongodb_uri=str())  # connection info for mongodb
+        self._stored.set_default(password_secret=str())
+
+    @property
+    def bind_address(self):
+        """The HTTP bind address used for the web interface"""
+        # TODO: come up with a more consistent way of doing this
+        addr = 'graylog-0.graylog-endpoints'
+        port = self.model.config['port']
+        return '{}:{}'.format(addr, port)
+
+    @property
+    def publish_uri(self):
+        """Web interface listen URI"""
+        return 'http://{}/'.format(self.bind_address)
+
+    @property
+    def external_uri(self):
+        """Ingress address of the Graylog application"""
+        ingress = str(self.model.get_binding('graylog').network.ingress_address)
+        port = self.model.config['port']
+        return 'http://{}:{}/'.format(ingress, port)
+
+    @property
+    def has_elasticsearch(self):
+        if self._stored.elasticsearch_uri:
+            return True
+        else:
+            return False
+
+    @property
+    def has_mongodb(self):
+        if self._stored.mongodb_uri:
+            return True
+        else:
+            return False
 
     def _on_config_changed(self, _):
-        current = self.model.config["thing"]
-        if current not in self._stored.things:
-            logger.debug("found a new thing: %r", current)
-            self._stored.things.append(current)
+        self._configure_pod()
+
+    def _on_update_status(self, _):
+        self._configure_pod()
+
+    def _on_stop(self, _):
+        self.unit.status = MaintenanceStatus('Pod is terminating.')
+
+    def _on_elasticsearch_relation_changed(self, event):
+        """Get the relation data from the relation and save to stored variable."""
+        # skip if unit is not leader
+        if not self.unit.is_leader():
+            return
+
+        data = event.relation.data[event.unit]
+        ingress_address = data.get('ingress-address')
+        port = data.get('port')
+        if ingress_address is None or port is None:
```

Comment:
Isn't it this case that address is always available in relation changed though it may not be available in relation joined .